Segmented inclusive_scan and exclusive_scan implementation

Sanac opened the pull request:
- Added segmented `inclusive_scan`. (see [#1338](https://github.com/STEllAR-GROUP/hpx/issues/1338))
- Added some simple tests

biddisco commented:
We had a long discussion about this today and my concern is that the first scan phase is done by each partition, but then the whold scan result is returned to the root partition where the merge/updte steps are performed and the final result vector is held on the root partition.

The reason for this was to do with iterators.

It seem that if the algorithm is called with a partitioned_vector as input, and a partitioned_vector as output (or in place), then those iterators are already present on the relevant partitions and the solution presented here is only useful if the output/dest iterator is actually a normal vector - and not a partitioned one.

@hkaiser - can you add a few lines to explain how to ensure that the merge/update phases will happen on the locality/partition that owns it, rather than back on the root.

hkaiser commented:
> can you add a few lines to explain how to ensure that the merge/update phases will happen on the locality/partition that owns it, rather than back on the root.

There is an example demonstrating how to work with the local parts of a partitioned vector only: https://github.com/STEllAR-GROUP/hpx/blob/master/examples/quickstart/partitioned_vector_spmd_foreach.cpp

Sanac commented:
- Added optimization to use locality of data if available
- In-place usage of the scan possible
- Changed tests so that changes are tested
- Also optimized tests so that less time is wasted doing setup and verification

Sanac commented:
- Refactoring of code in preparation for exclusive scan
- Code can be reviewed

Sanac commented:
- Added some comments in code
- Added exclusive scan with tests to PR

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -0,0 +1,792 @@
+
+                if(beg_in != end_in) {
+                    std::uint64_t in_id = traits_in::get_id(sit_in).get_msb();
+                    std::uint64_t out_id = traits_in::get_id(sit_in).get_msb();
```
Shouldn't this compare the whole id and not only the `msb`?

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -0,0 +1,792 @@
+                    std::size_t in_dist = std::distance(beg_in, end_in);
+                    std::size_t out_dist = std::distance(beg_out, end_out);
+
+                    if (in_id != out_id || in_dist != out_dist) {
```
Minor nitpick: there is no need to invoke `std::distance` if the ids are different.

hkaiser commented on hpx/parallel/segmented_algorithms/detail/scan.hpp:
```diff
@@ -0,0 +1,792 @@
+            typename T, typename Op>
+        static typename util::detail::algorithm_result<ExPolicy, OutIter>::type
+        segmented_scan_par(ExPolicy const& policy, SegIter first,
+            SegIter last, OutIter dest, T init, Op && op, std::true_type)
```
Once `init` has been copied at the uppermost API level it could be passed down by `const&`.

hkaiser commented on hpx/parallel/segmented_algorithms/exclusive_scan.hpp:
```diff
@@ -0,0 +1,282 @@
+
+                // use first element to save last T of scan
+                return result::get(
+                    dataflow([=, &op](vector_type r) {
```
Is capturing `op` by reference safe in this context?

Sanac replied:
Lambdas are still new to me. Thanks for your comment.

hkaiser commented on hpx/parallel/segmented_algorithms/inclusive_scan.hpp:
```diff
@@ -0,0 +1,278 @@
+                }
+ 
+                return result::get(
+                    dataflow([=, &op](vector_type r) {
```
Same here, is capturing `op` by reference safe?

hkaiser commented on tests/unit/component/partitioned_vector_exclusive_scan.cpp:
```diff
@@ -0,0 +1,360 @@
+
+int main()
+{
+    std::vector<hpx::id_type> localities = hpx::find_remote_localities();
```
Is this test supposed to run in distributed?

Sanac replied:
Yes those should be distributed to test the version of the algorithm that use the locality of the data. (i.e. scan is performed on partition that owns the data instead of the root partition)

hkaiser commented on tests/unit/component/partitioned_vector_inclusive_scan.cpp:
```diff
@@ -0,0 +1,362 @@
+
+int main()
+{
+    std::vector<hpx::id_type> localities = hpx::find_remote_localities();
```
Same here: should this run in distributed, i.e. should it use the locality ids?

Sanac replied:
Same as the other test.

hkaiser replied:
I now understand, the find_remote_localities() call is not necessary here as it is done again in the called functions. Alternatively you may pass `localities` to the actual tests.

Sanac commented:
- Fixed comparison of `msb` to locality id
- Invocation of `std::distance` only if ids are the same
- Passed down `init` by `const&`
- Capturing of `op` done by value

Sanac commented:
- Pass localities via parameter in the tests
